File: src/modules/db1/mgmt_nonce.c
/* db1/mgmt_nonce.c: the management challenge rows and the revocation HWM.
 *
 * Lifted out of server/server_mgmt_status.c. What stayed there is the peer
 * certificate, the signature check, and the decision that a request is valid;
 * what came here is every statement that touched the database.
 *
 * The nonce is 32 bytes and arrives as hex (see the header). This file is the
 * only place that converts, and it refuses anything that is not exactly 64
 * lowercase hex characters rather than binding a short value -- a truncated
 * nonce would look up a different challenge.
 */
#include "mgmt_nonce.h"
#include "db1_internal.h"

#include <sqlite3.h>
#include <string.h>

static int nonce_bytes(const char *hex, unsigned char out[32])
{
   if (!hex || strlen(hex) != 64)
      return -1;
   for (int i = 0; i < 32; i++)
   {
      int hi = hex[i * 2], lo = hex[i * 2 + 1];
      hi = (hi >= '0' && hi <= '9') ? hi - '0' : (hi >= 'a' && hi <= 'f') ? hi - 'a' + 10 : -1;
      lo = (lo >= '0' && lo <= '9') ? lo - '0' : (lo >= 'a' && lo <= 'f') ? lo - 'a' + 10 : -1;
      if (hi < 0 || lo < 0)
         return -1;
      out[i] = (unsigned char)((hi << 4) | lo);
   }
   return 0;
}

static int text_eq(sqlite3_stmt *s, int col, const char *expected)
{
   const unsigned char *v = sqlite3_column_text(s, col);
   return v && expected && strcmp((const char *)v, expected) == 0;
}

int db1_mgmt_nonce_clear(void)
{
   sqlite3 *db = db1_conn();
   return db && sqlite3_exec(db, "DELETE FROM server_mgmt_nonce", NULL, NULL, NULL) == SQLITE_OK
              ? 0
              : -1;
}

int db1_mgmt_nonce_issue(const db1_mgmt_nonce_issue_t *in)
{
   unsigned char nonce[32];
   if (!in || nonce_bytes(in->nonce, nonce) != 0)
      return DB1_MGMT_NONCE_INVALID;
   sqlite3 *db = db1_conn();
   if (!db)
      return DB1_MGMT_NONCE_STORAGE;
   if (db1_txn_begin(db, "BEGIN IMMEDIATE") != 0)
      return DB1_MGMT_NONCE_STORAGE;

   int result = DB1_MGMT_NONCE_STORAGE;
   sqlite3_stmt *q = NULL;
   if (sqlite3_prepare_v2(db, "DELETE FROM server_mgmt_nonce WHERE expires_at<?1", -1, &q, NULL) !=
       SQLITE_OK)
      goto done;
   sqlite3_bind_int64(q, 1, (sqlite3_int64)in->now);
   if (sqlite3_step(q) != SQLITE_DONE)
      goto done;
   sqlite3_finalize(q);
   q = NULL;
   if (sqlite3_prepare_v2(db, "SELECT count(*) FROM server_mgmt_nonce", -1, &q, NULL) !=
           SQLITE_OK ||
       sqlite3_step(q) != SQLITE_ROW)
      goto done;
   int count = sqlite3_column_int(q, 0);
   sqlite3_finalize(q);
   q = NULL;
   if (count >= DB1_MGMT_NONCE_CAP)
   {
      result = DB1_MGMT_NONCE_SATURATED;
      goto commit;
   }
   static const char sql[] =
       "INSERT INTO server_mgmt_nonce(nonce,peer_issuer,peer_serial_norm,peer_fingerprint,"
       "channel_binding,target_server_id,purpose,expires_at) VALUES(?1,?2,?3,?4,?5,?6,"
       "?7,?8)";
   if (sqlite3_prepare_v2(db, sql, -1, &q, NULL) != SQLITE_OK)
      goto done;
   sqlite3_bind_blob(q, 1, nonce, 32, SQLITE_TRANSIENT);
   sqlite3_bind_text(q, 2, in->peer_issuer, -1, SQLITE_TRANSIENT);
   sqlite3_bind_text(q, 3, in->peer_serial_norm, -1, SQLITE_TRANSIENT);
   sqlite3_bind_text(q, 4, in->peer_fingerprint, -1, SQLITE_TRANSIENT);
   sqlite3_bind_text(q, 5, in->channel_binding, -1, SQLITE_TRANSIENT);
   sqlite3_bind_text(q, 6, in->target_server_id, -1, SQLITE_TRANSIENT);
   sqlite3_bind_text(q, 7, in->purpose, -1, SQLITE_TRANSIENT);
   sqlite3_bind_int64(q, 8, (sqlite3_int64)(in->now + DB1_MGMT_NONCE_TTL));
   if (sqlite3_step(q) != SQLITE_DONE)
      goto done;
   result = DB1_MGMT_NONCE_OK;
commit:
   sqlite3_finalize(q);
   return db1_txn_end(db, "COMMIT") == 0 ? result : DB1_MGMT_NONCE_STORAGE;
done:
   sqlite3_finalize(q);
   db1_txn_end(db, "ROLLBACK");
   return DB1_MGMT_NONCE_STORAGE;
}

int db1_mgmt_nonce_consume(const db1_mgmt_nonce_consume_t *in)
{
   unsigned char nonce[32];
   if (!in || nonce_bytes(in->nonce, nonce) != 0)
      return DB1_MGMT_NONCE_INVALID;
   sqlite3 *db = db1_conn();
   if (!db)
      return DB1_MGMT_NONCE_STORAGE;
   if (db1_txn_begin(db, "BEGIN IMMEDIATE") != 0)
      return DB1_MGMT_NONCE_STORAGE;

   int result = DB1_MGMT_NONCE_STORAGE;
   sqlite3_stmt *q = NULL;
   static const char lookup[] =
       "SELECT peer_issuer,peer_serial_norm,peer_fingerprint,channel_binding,target_server_id,"
       "purpose,expires_at FROM server_mgmt_nonce WHERE nonce=?1";
   if (sqlite3_prepare_v2(db, lookup, -1, &q, NULL) != SQLITE_OK)
      goto rollback;
   sqlite3_bind_blob(q, 1, nonce, 32, SQLITE_TRANSIENT);
   if (sqlite3_step(q) != SQLITE_ROW)
   {
      result = DB1_MGMT_NONCE_NOT_FOUND;
      goto commit;
   }
   int64_t exp = sqlite3_column_int64(q, 6);
   int bound = text_eq(q, 0, in->peer_issuer) && text_eq(q, 1, in->peer_serial_norm) &&
               text_eq(q, 2, in->peer_fingerprint) && text_eq(q, 3, in->channel_binding) &&
               text_eq(q, 4, in->target_server_id) && text_eq(q, 5, in->purpose);
   sqlite3_finalize(q);
   q = NULL;
   /* Deleted whatever the verdict is: a challenge that was looked at is spent,
      or a wrong answer would be a free retry. */
   if (sqlite3_prepare_v2(db, "DELETE FROM server_mgmt_nonce WHERE nonce=?1", -1, &q, NULL) !=
       SQLITE_OK)
      goto rollback;
   sqlite3_bind_blob(q, 1, nonce, 32, SQLITE_TRANSIENT);
   if (sqlite3_step(q) != SQLITE_DONE || sqlite3_changes(db) != 1)
      goto rollback;
   sqlite3_finalize(q);
   q = NULL;
   if (!bound)
      result = DB1_MGMT_NONCE_MISMATCH;
   else if (exp < 0 || in->now > exp)
      result = DB1_MGMT_NONCE_EXPIRED;
   else if (!in->valid)
      result = DB1_MGMT_NONCE_INVALID;
   else
   {
      if (sqlite3_prepare_v2(db, "SELECT generation FROM server_mgmt_status_hwm WHERE singleton=1",
                             -1, &q, NULL) != SQLITE_OK ||
          sqlite3_step(q) != SQLITE_ROW)
         goto rollback;
      int64_t hwm = sqlite3_column_int64(q, 0);
      sqlite3_finalize(q);
      q = NULL;
      if (in->revocation_generation < hwm)
         result = DB1_MGMT_NONCE_ROLLBACK;
      else if (sqlite3_prepare_v2(db,
                                  "UPDATE server_mgmt_status_hwm SET generation=max(generation,?1) "
                                  "WHERE singleton=1",
                                  -1, &q, NULL) != SQLITE_OK)
         goto rollback;
      else
      {
         sqlite3_bind_int64(q, 1, (sqlite3_int64)in->revocation_generation);
         if (sqlite3_step(q) != SQLITE_DONE || sqlite3_changes(db) != 1)
            goto rollback;
         result = DB1_MGMT_NONCE_OK;
      }
   }
commit:
   sqlite3_finalize(q);
   return db1_txn_end(db, "COMMIT") == 0 ? result : DB1_MGMT_NONCE_STORAGE;
rollback:
   sqlite3_finalize(q);
   db1_txn_end(db, "ROLLBACK");
   return DB1_MGMT_NONCE_STORAGE;
}

int db1_mgmt_status_hwm_read(int64_t *generation)
{
   sqlite3 *db = db1_conn();
   sqlite3_stmt *q = NULL;
   if (!db || !generation ||
       sqlite3_prepare_v2(db, "SELECT generation FROM server_mgmt_status_hwm WHERE singleton=1", -1,
                          &q, NULL) != SQLITE_OK)
      return -1;
   int rc = sqlite3_step(q) == SQLITE_ROW ? 0 : -1;
   if (rc == 0)
      *generation = sqlite3_column_int64(q, 0);
   sqlite3_finalize(q);
   return rc;
}

int db1_mgmt_status_hwm_set(int64_t generation)
{
   sqlite3 *db = db1_conn();
   sqlite3_stmt *q = NULL;
   if (!db || generation < 0 ||
       sqlite3_prepare_v2(db,
                          "UPDATE server_mgmt_status_hwm SET generation=?1 "
                          "WHERE singleton=1 AND generation<=?1",
                          -1, &q, NULL) != SQLITE_OK)
      return -1;
   sqlite3_bind_int64(q, 1, (sqlite3_int64)generation);
   int rc = sqlite3_step(q) == SQLITE_DONE && sqlite3_changes(db) == 1 ? 0 : -1;
   sqlite3_finalize(q);
   return rc;
}

File: server-go/modules/aimee/families/mgmt_nonce.go
package families

import (
	"context"
	"encoding/hex"
	"errors"

	store "github.com/JBailes/aimee/server-go/modules/aimee"
)

// The management challenge postgres: a short-lived nonce per outstanding
// challenge, plus the revocation high-water mark a successful consume advances.
const (
	EventMgmtNonce uint32 = 11794
	StageMgmtNonce uint32 = 18

	opMgmtNonceClear    uint32 = 1
	opMgmtNonceIssue    uint32 = 2
	opMgmtNonceConsume  uint32 = 3
	opMgmtStatusHWMRead uint32 = 4
	opMgmtStatusHWMSet  uint32 = 5
)

// Result codes from mgmt_nonce.h. As elsewhere in this module these ride in the reply
// field, not the wire status.
const (
	nonceOK        = 0
	nonceNotFound  = 1
	nonceMismatch  = 2
	nonceExpired   = 3
	nonceRollback  = 4
	nonceInvalid   = 5
	nonceSaturated = 6
	nonceStorage   = 7
)

// nonceCap and nonceTTL are DB1_MGMT_NONCE_CAP and _TTL. The TTL is in seconds
// and is added to the caller's clock, so the store never reads its own.
const (
	nonceCap = 128
	nonceTTL = 15
)

const (
	nonceClearSQL = `DELETE FROM server_mgmt_nonce`

	nonceSweepSQL = `DELETE FROM server_mgmt_nonce WHERE expires_at < $1`

	nonceCountSQL = `SELECT count(*) FROM server_mgmt_nonce`

	nonceInsertSQL = `INSERT INTO server_mgmt_nonce
	                      (nonce, peer_issuer, peer_serial_norm, peer_fingerprint,
	                       channel_binding, target_server_id, purpose, expires_at)
	                  VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`

	nonceLookupSQL = `SELECT peer_issuer, peer_serial_norm, peer_fingerprint, channel_binding,
	                         target_server_id, purpose, expires_at
	                    FROM server_mgmt_nonce
	                   WHERE nonce = $1`

	nonceDeleteSQL = `DELETE FROM server_mgmt_nonce WHERE nonce = $1`

	hwmReadSQL = `SELECT generation FROM server_mgmt_status_hwm WHERE singleton`

	hwmBumpSQL = `UPDATE server_mgmt_status_hwm
	                 SET generation = greatest(generation, $1)
	               WHERE singleton`

	hwmSetSQL = `UPDATE server_mgmt_status_hwm
	                SET generation = $1
	              WHERE singleton AND generation <= $1`
)

// decodeNonce converts the wire's hex to the store's 32 bytes.
//
// Exactly 64 LOWERCASE hex characters, refused otherwise rather than decoded
// leniently: a truncated nonce would look up a different challenge, and Go's
// hex decoder would happily accept the uppercase spelling the C never did.
func decodeNonce(field string) ([]byte, bool) {
	if !lowercaseHex(field, 64, 64) {
		return nil, false
	}
	raw, err := hex.DecodeString(field)
	if err != nil {
		return nil, false
	}
	return raw, true
}

// nonceResult is the single-cell reply every nonce operation with an outcome
// answers with.
func nonceResult(code int) (uint32, []string, error) {
	return store.StatusOK, []string{store.Itoa(code)}, nil
}

// nonceClear is op 1: drop every outstanding challenge.
func nonceClear(ctx context.Context, q store.Queryer, _ []string) (uint32, []string, error) {
	if _, err := q.Exec(ctx, nonceClearSQL); err != nil {
		return 0, nil, err
	}
	return store.StatusOK, nil, nil
}

// nonceIssue is op 2: record a challenge the server just handed out.
//
// Sweep, then count, then insert -- in that order, so a store full of expired
// challenges is not reported as saturated. Saturation COMMITS the sweep for the
// same reason the jti stores do: discarding it would mean a saturated store can
// never drain.
func nonceIssue(ctx context.Context, db store.DB, f []string) (uint32, []string, error) {
	nonce, ok := decodeNonce(f[0])
	if !ok {
		return nonceResult(nonceInvalid)
	}
	now, okNow := store.Atoi64(f[7])
	if !okNow {
		return nonceResult(nonceInvalid)
	}

	tx, err := db.Begin(ctx)
	if err != nil {
		return nonceResult(nonceStorage)
	}
	defer func() { _ = tx.Rollback(ctx) }()

	if _, err := tx.Exec(ctx, nonceSweepSQL, now); err != nil {
		return nonceResult(nonceStorage)
	}
	var live int64
	if err := tx.QueryRow(ctx, nonceCountSQL).Scan(&live); err != nil {
		return nonceResult(nonceStorage)
	}
	if live >= nonceCap {
		if err := tx.Commit(ctx); err != nil {
			return nonceResult(nonceStorage)
		}
		return nonceResult(nonceSaturated)
	}
	// The expiry is the caller's clock plus the TTL. The store never consults
	// its own clock, so a caller and the store cannot disagree about now.
	if _, err := tx.Exec(ctx, nonceInsertSQL, nonce, f[1], f[2], f[3], f[4], f[5], f[6],
		now+nonceTTL); err != nil {
		return nonceResult(nonceStorage)
	}
	if err := tx.Commit(ctx); err != nil {
		return nonceResult(nonceStorage)
	}
	return nonceResult(nonceOK)
}

// nonceConsume is op 3: spend a challenge and, if everything checks out,
// advance the revocation high-water mark.
//
// The challenge is deleted whatever the verdict, which is the point: a
// challenge that was looked at is spent, and leaving a mismatched one in place
// would make a wrong answer a free retry.
func nonceConsume(ctx context.Context, db store.DB, f []string) (uint32, []string, error) {
	nonce, ok := decodeNonce(f[0])
	if !ok {
		return nonceResult(nonceInvalid)
	}
	now, okNow := store.Atoi64(f[7])
	generation, okGen := store.Atoi64(f[8])
	valid, okValid := store.Atoi(f[9])
	if !okNow || !okGen || !okValid {
		return nonceResult(nonceInvalid)
	}

	tx, err := db.Begin(ctx)
	if err != nil {
		return nonceResult(nonceStorage)
	}
	defer func() { _ = tx.Rollback(ctx) }()

	var issuer, serial, fingerprint, binding, target, purpose string
	var expiresAt int64
	err = tx.QueryRow(ctx, nonceLookupSQL, nonce).Scan(&issuer, &serial, &fingerprint,
		&binding, &target, &purpose, &expiresAt)
	switch {
	case store.IsNoRows(err):
		// Nothing to spend. Committing rather than rolling back keeps this the
		// same shape as every other path, and there is nothing to undo.
		if err := tx.Commit(ctx); err != nil {
			return nonceResult(nonceStorage)
		}
		return nonceResult(nonceNotFound)
	case err != nil:
		return nonceResult(nonceStorage)
	}

	bound := issuer == f[1] && serial == f[2] && fingerprint == f[3] &&
		binding == f[4] && target == f[5] && purpose == f[6]

	tag, err := tx.Exec(ctx, nonceDeleteSQL, nonce)
	if err != nil {
		return nonceResult(nonceStorage)
	}
	if tag.RowsAffected() != 1 {
		// The row was there a statement ago. If it is not exactly one now,
		// something else is writing this table and the outcome is not safe to
		// report as a verdict.
		return nonceResult(nonceStorage)
	}

	switch {
	case !bound:
		return commitWith(ctx, tx, nonceMismatch)
	case expiresAt < 0 || now > expiresAt:
		return commitWith(ctx, tx, nonceExpired)
	case valid == 0:
		return commitWith(ctx, tx, nonceInvalid)
	}

	var hwm int64
	if err := tx.QueryRow(ctx, hwmReadSQL).Scan(&hwm); err != nil {
		return nonceResult(nonceStorage)
	}
	if generation < hwm {
		// A replayed status document: it carries a revocation generation older
		// than one already seen, so accepting it would roll the peer back to a
		// state where a revoked credential was still live.
		return commitWith(ctx, tx, nonceRollback)
	}
	tag, err = tx.Exec(ctx, hwmBumpSQL, generation)
	if err != nil {
		return nonceResult(nonceStorage)
	}
	if tag.RowsAffected() != 1 {
		return nonceResult(nonceStorage)
	}
	return commitWith(ctx, tx, nonceOK)
}

// commitWith commits the spend and reports the verdict. The DELETE has to
// persist for every one of these outcomes, so the verdict never decides whether
// to commit -- only what to say.
func commitWith(ctx context.Context, tx store.Tx, code int) (uint32, []string, error) {
	if err := tx.Commit(ctx); err != nil {
		return nonceResult(nonceStorage)
	}
	return nonceResult(code)
}

// hwmRead is op 4: the current revocation high-water mark.
func hwmRead(ctx context.Context, q store.Queryer, _ []string) (uint32, []string, error) {
	var generation int64
	switch err := q.QueryRow(ctx, hwmReadSQL).Scan(&generation); {
	case store.IsNoRows(err):
		// The seed row is part of the schema, so its absence is a broken store
		// rather than a cold one.
		return 0, nil, errors.New("aimee: the status high-water mark row is missing")
	case err != nil:
		return 0, nil, err
	}
	return store.StatusOK, []string{store.I64toa(generation)}, nil
}

// hwmSet is op 5: move the mark forward.
//
// Monotonic: the UPDATE carries `generation <= $1`, so a lower value matches no
// row and is refused rather than silently winding the mark back.
func hwmSet(ctx context.Context, q store.Queryer, f []string) (uint32, []string, error) {
	generation, ok := store.Atoi64(f[0])
	if !ok || generation < 0 {
		return store.StatusInvalid, nil, nil
	}
	tag, err := q.Exec(ctx, hwmSetSQL, generation)
	if err != nil {
		return 0, nil, err
	}
	if tag.RowsAffected() != 1 {
		// Either the row is missing or the caller tried to move the mark
		// backwards. Both are failures rather than quiet successes.
		return store.StatusFailed, nil, nil
	}
	return store.StatusOK, nil, nil
}

// MgmtNonce is the family, ready to be bound to kind 11794.
var MgmtNonce = store.Family{
	Name:  "mgmt_nonce",
	Event: EventMgmtNonce,
	Stage: StageMgmtNonce,
	Ops: map[uint32]store.Op{
		opMgmtNonceClear:    {Name: "mgmt_nonce_clear", Args: 0, Tx: true, Run: nonceClear},
		opMgmtNonceIssue:    {Name: "mgmt_nonce_issue", Args: 8, RunDB: nonceIssue},
		opMgmtNonceConsume:  {Name: "mgmt_nonce_consume", Args: 10, RunDB: nonceConsume},
		opMgmtStatusHWMRead: {Name: "mgmt_status_hwm_read", Args: 0, Run: hwmRead},
		opMgmtStatusHWMSet:  {Name: "mgmt_status_hwm_set", Args: 1, Tx: true, Run: hwmSet},
	},
}
